include/osquery/registry.h

```diff
@@ -15,14 +15,294 @@
 #include <unordered_map>
 #include <utility>
 
+#include <functional>
+#include <map>
+#include <mutex>
+#include <vector>
+
+#include <boost/noncopyable.hpp>
+
 #include <osquery/database/results.h>
 #include <osquery/logger.h>
 
-#include "osquery/registry/init_registry.h"
-#include "osquery/registry/singleton.h"
-
 namespace osquery {
 
+typedef std::map<std::string, std::string> RouteInfo;
+typedef std::map<std::string, RouteInfo> RegistryRoutes;
+typedef std::map<std::string, RegistryRoutes> RegistryBroadcast;
+
+typedef std::map<std::string, std::string> PluginRequest;
+typedef std::map<std::string, std::string> PluginReponse;
+
+template <class RegistryType>
+class NewRegistry {
+ protected:
+  typedef std::shared_ptr<RegistryType> RegistryTypeRef;
+  virtual void type() {}
+
+ public:
+  template <class Item>
+  Status add(const std::string& item_name) {
+    if (items_.count(item_name) > 0) {
+      return Status(1, "Duplicate registry item exists: " + item_name);
+    }
+    auto item = reinterpret_cast<RegistryType*>(new Item());
+
+    auto status = item->setUp();
+    if (!status.ok()) {
+      return status;
+    }
+
+    item->setName(item_name);
+    std::shared_ptr<RegistryType> shared_item(item);
+    items_[item_name] = shared_item;
+    return Status(0, "OK");
+  }
+
+  RegistryTypeRef get(const std::string& item_name) {
+    if (items_.count(item_name) > 0) {
+      return items_[item_name];
+    }
+
+    // Since this is a reference request, a domain error is thrown if the
+    // item does not exist in the registry.
+    throw std::domain_error("Cannot find registry item: " + item_name);
+  }
+
+  RegistryRoutes getRoutes() {
+    RegistryRoutes route_table;
+    for (const auto& item : items_) {
+      route_table[item.first] = item.second->routeInfo();
+    }
+    return route_table;
+  }
+
+  Status call(const std::string& item_name,
+              const PluginRequest& request,
+              PluginRequest& response) {
+    if (items_.count(item_name) > 0) {
+      return items_[item_name]->call(request, response);
+    }
+    return Status(1, "Cannot call registry item: " + item_name);
```

Comment:
Should this method call `get`? Are we okay with `get` throwing an exception on error, but this returning a `Status`?